src/protocol/boolean/bit_decomposition.rs

```diff
@@ -38,49 +40,170 @@ impl BitDecomposition {
+
+        // Step 4.2. Compute [g]_B = (f_i - c_i) [q]_p + c_i
+        let mut g_bin: Vec<G> = Vec::with_capacity(usize::try_from(el).unwrap());
+        for bit_index in 0..el {
+            // these are single bits, so let's make them bools
+            let f_i: bool = (f_int >> bit_index) & 1 == 1;
+            let c_i: bool = (c.as_u128() >> bit_index) & 1 == 1;
+            // g_i can either be c_i (known to be either 0, 1), or either [q]_p, ¬[q]_p
+            // using an enum here because we're going to do something cleaver in bitwise add
+            let g_i: G = match (f_i, c_i) {
+                // Case where f_i - c_i == 0
+                (false, false) | (true, true) => match c_i {
+                    true => G::One,
+                    false => G::Zero,
+                },
+                // Case where f_i - c_i = 1
+                (true, false) => G::Q,
+                // Case where f_i - c_i = -1
+                (false, true) => G::NotQ,
+                // this is what g_i should actually be here, so we can use this down the road
+                // let g_i_foo = S::share_known_value(&ctx, F::ONE) - &q_p.clone();
+            };
+            g_bin.push(g_i);
+            // let f_i_minus_c_i: u64 = f_i - c_i;
+            // let g_prime: S = q_p.clone().into() * f_i_minus_c_i;
+            // // c_i ∈ {0, 1}, so F::truncate_from will work
+            // debug_assert!(c_i <= 1_u128);
+            // g_bin.push(S::share_known_value(&ctx, F::truncate_from(c_i)) + &g_prime);
+        }
+
+        // Step 5. Compute BitwiseSum([r]_B, [g]_B)
+        let mut h: Vec<S> = Vec::with_capacity(usize::try_from(el).unwrap());
+        let one_minus_q_p = S::share_known_value(&ctx, F::ONE) - &q_p.clone();
+        let (mut last_carry, result_bit, mut last_carry_known_to_be_zero) = match g_bin[0] {
+            G::Zero => (S::ZERO, r.b_b[0].clone(), true),
+            G::One => (r.b_b[0].clone(), S::share_known_value(&ctx, F::ONE), false),
+            G::Q => (
+                q_p.multiply(&r.b_b[0], ctx.narrow(&Step::AddGtoR), record_id)
+                    .await?,
+                q_p.clone() + &r.b_b[0],
+                false,
+            ),
+            G::NotQ => (
+                one_minus_q_p
+                    .clone()
+                    .multiply(&r.b_b[0], ctx.narrow(&Step::AddGtoR), record_id)
+                    .await?,
+                one_minus_q_p.clone() + &r.b_b[0],
+                false,
+            ),
+        };
+        h.push(result_bit);
 
-        // Step 8, 9. [g_i] = [q] * f_i
-        // Step 10. [h]_B = [d + g]_B, where [h]_B = ([h]_0,...[h]_(el+1))
-        // Step 11. [a]_B = ([h]_0,...[h]_(el-1))
-        let a_b =
-            maybe_add_constant_mod2l(ctx.narrow(&Step::AddDtoG), record_id, &d_b, x, &q_p).await?;
+        for (bit_index, bit) in r.b_b.iter().enumerate().skip(1) {
+            let mult_result = if last_carry_known_to_be_zero {
+                // TODO: this makes me sad
```

Comment:
what is going on here?

Reply:
This is code that Ben and I wrote to get around the fact that we can't have holes in the sequence record indices.  That is, if you multiply record N, you also have to multiply records 0 through N-1 or the infrastructure stalls.

Reply:
yea, I have to do even more of this with this PR.

Reply:
this specific line is outdated, and Alex and I have talked about it, so resolving this comment.